### PR TITLE
Custom mustache delimiters

### PR DESCRIPTION
**Pull request by alabiaga**
Fixes: #12994 

- [x] add support for custom delimiters in script tag defined templates.
- [x] add support for custom delimiters in template tag defined templates.

Allow for defining custom delimiters via data-custom-delimiters attribute.

e.g.

    <amp-list ..>
	  <template type="amp-mustache" data-custom-delimiters="#-,-#">
         ...
      </template>
	</amp-list>

    or

    <amp-list ..>
	  <script type="text/plain" template="amp-mustache"
          data-custom-delimiters="#-,-#">
         ...
      </script>
	</amp-list>

### CODE REVIEW

**Comment by alabiaga** [comment]
cc\ @morsssss

File: extensions/amp-mustache/0.1/amp-mustache.js [hunk 1]
```diff
@@ -61,30 +61,66 @@ export class AmpMustache extends AMP.BaseTemplate {
...
+     * Used to encode html entities in delimiters.
+     * @private {!Element}
+     */
+    this.textArea_ = document.createElement('textarea');
+
     /** @private @const {string} */
     this.template_ = this.initTemplateString_();
 
-    mustache.parse(this.template_, /* tags */ undefined);
+    mustache.parse(this.template_, /* tags */ this.getDelimiters_());
```
Comment (jridgewell) [inline extensions/amp-mustache/0.1/amp-mustache.js]: Should we be adding new features to the 0.1 mustache component?

File: extensions/amp-mustache/0.2/amp-mustache.js [hunk 2]
```diff
@@ -55,30 +56,67 @@ export class AmpMustache extends AMP.BaseTemplate {
     /** @private @const {!JsonObject} */
     this.nestedTemplates_ = dict();
 
+    /**
+     * Used to encode html entities in delimiters.
+     * @private {!Element}
+     */
+    this.textArea_ = document.createElement('textarea');
```
Comment (jridgewell) [inline extensions/amp-mustache/0.2/amp-mustache.js]: There's no reason to hold this as a property.

File: extensions/amp-mustache/0.2/amp-mustache.js [hunk 3]
```diff
@@ -55,30 +56,67 @@ export class AmpMustache extends AMP.BaseTemplate {
...
+   */
+  getDelimiters_() {
+    const CUSTOM_DELIMITERS_ATTR = 'data-custom-delimiters';
+    let delimiters = null;
+    const {element} = this;
+    if (element.hasAttribute(CUSTOM_DELIMITERS_ATTR)) {
+      const delimitersStr = element.getAttribute(CUSTOM_DELIMITERS_ATTR);
+      userAssert(delimitersStr.split(',').length == 2,
+          'Beginning and ending delimiter is required: %s.', element);
+      delimiters = delimitersStr.split(',');
```
Comment (jridgewell) [inline extensions/amp-mustache/0.2/amp-mustache.js]: Nit: do this above, avoid the double split.

File: extensions/amp-mustache/0.2/test/test-amp-mustache.js [hunk 4]
```diff
@@ -110,6 +110,41 @@ describes.repeated('amp-mustache 0.2', {
         .equal('<a href="javascript:alert(0)">abc</a>');
   });
 
+  describe('custom delimiters', () => {
+    it('should require beginning and ending delimiter', () => {
+      templateElement.setAttribute('data-custom-delimiters', '<%');
+      innerHtmlSetup(
```
Comment (jridgewell) [inline extensions/amp-mustache/0.2/test/test-amp-mustache.js]: I can't comment on the appropriate line: https://github.com/ampproject/amphtml/blob/7981c6743bc5712e18ec281b185a54cba2dd7a95/extensions/amp-mustache/0.1/test/test-amp-mustache.js#L69

I believe is incorrect, you need to insert text into the script element using `innerHTML`.

**Comment by morsssss** [comment]
Thanks! It was fun also to see how you did this :)

File: extensions/amp-mustache/0.2/amp-mustache.js [hunk 5]
```diff
@@ -58,27 +59,59 @@ export class AmpMustache extends AMP.BaseTemplate {
...
+      // If using a template encode any html entities used in a delimiter.
+      // This is required as the template will also contain decoded delimiters
+      // prior to being parsed by mustache.
+      if (element.tagName == 'TEMPLATE') {
+        const textArea = document.createElement('textarea');
+        for (let i = 0; i < delimiters.length; i++) {
+          const delimiter = delimiters[i];
+          textArea.textContent = delimiter;
+          delimiters[i] = textArea./*OK*/innerHTML;
+        }
```
Comment (choumx) [inline extensions/amp-mustache/0.2/amp-mustache.js]: Nit: It's not obvious what this code is doing here, so a well-named helper function may help readability:

```js
if (element.tagName == 'TEMPLATE') {
  delimiters = htmlEncode(delimiters);
}
```

File: extensions/amp-mustache/0.2/test/test-amp-mustache.js [hunk 6]
```diff
@@ -64,11 +64,7 @@ describes.repeated('amp-mustache 0.2', {
       }
     };
     innerHtmlSetup = html => {
-      if (isTemplateType) {
-        templateElement./*OK*/innerHTML = html;
-      } else if (isTemplateTypeScript) {
-        templateElement.textContent = html;
-      }
+      templateElement./*OK*/innerHTML = html;
```
[the remark this answers is not in the document]
Comment (alabiaga) [reply extensions/amp-mustache/0.2/test/test-amp-mustache.js]: no difference apparently. When I first created the PR, I thought that setting the contents of the script or template tag differed when it came to the API but this doesn't seem to be the case at all, as innerHTML and textContent in this case achieves the same goal. Though I recall this not working initially and having to make these changes for a reason.

**Comment by alabiaga** [comment]
@choumx PTAL when you get a chance. Thanks

File: extensions/amp-mustache/0.2/amp-mustache.js [hunk 7]
```diff
@@ -58,27 +63,80 @@ export class AmpMustache extends AMP.BaseTemplate {
...
+    return delimiters;
+  }
+
+  /**
+   * @param {Array<string>} delimiters
+   * @return {boolean}
+   * @private
+   */
+  validDelimiters_(delimiters) {
+    return delimiters.some(delimiter => {
```
Comment (jridgewell) [inline extensions/amp-mustache/0.2/amp-mustache.js]: This would pass with one valid delimiter.

File: extensions/amp-mustache/0.2/amp-mustache.js [hunk 8]
```diff
@@ -58,27 +63,80 @@ export class AmpMustache extends AMP.BaseTemplate {
...
+      return delimiter.trim() !== '' && delimiter !== '=';
```
Comment (jridgewell) [inline extensions/amp-mustache/0.2/amp-mustache.js]: Does mustache throw an error?

File: extensions/amp-mustache/0.2/amp-mustache.js (hunk 7 above)
Comment (alabiaga) [reply extensions/amp-mustache/0.2/amp-mustache.js]: ah good catch. added an additional test to test this. thanks.

File: extensions/amp-mustache/0.2/amp-mustache.js (hunk 8 above)
Comment (alabiaga) [reply extensions/amp-mustache/0.2/amp-mustache.js]: it does but the console error would be the following: `Error: Unclosed tag at 8​​​..`

File: examples/amp-list.amp.html [hunk 9]
```diff
@@ -52,8 +52,8 @@
   The list content is rendered via an [amp-mustache template](https://www.ampproject.org/docs/reference/components/amp-mustache). The template can be specified either by id, or by using a nested element.
   -->
   <amp-list id="list1" width="auto" height="100" layout="fixed-height" src="list.example.json" class="m1" reset-on-refresh>
-    <template type="amp-mustache" id="amp-template-id">
-      <div><a href="{{url}}">{{title}}</a></div>
+    <template type="amp-mustache" id="amp-template-id" data-custom-delimiters="<%,%>">
```
[the remark this answers is not in the document]
Comment (choumx) [reply examples/amp-list.amp.html]: Any thoughts on this suggestion?

File: extensions/amp-mustache/0.2/amp-mustache.js (hunk 8 above)
Comment (jridgewell) [reply extensions/amp-mustache/0.2/amp-mustache.js]: Ok. Does mustache throw errors in any other circumstance?

**Comment by choumx** [comment]
This would be a nice and short design review topic if you're interested in honing those presentation skills. :)

**Comment by choumx** [comment]
Dug up this old snippet, just FYI: https://github.com/ampproject/amphtml/blob/dd0708dab7ada5b74f00f018fbbe8b60e9833b75/third_party/mustache/mustache.js#L209-L216

**Comment by alabiaga** [comment]
> This would be a nice and short design review topic if you're interested in honing those presentation skills. :)

Sure thing. I'll sign up for the next one.

File: examples/amp-list.amp.html [hunk 10]
```diff
@@ -75,6 +83,24 @@
...
 
+  <!-- ## Using a script template with custom mustache delimiters. -->
+  <amp-list width="auto" height="100" layout="fixed-height" src="comments.example.json">
+    <script type="text/plain" template="amp-mustache" custom-delimiters="#-,-#">
+      <table>
+        <tr>
+          <td>Comment:</td>
+          <td>#-content-#</td>
+        </tr>
+        #-#replies-#
```
Comment (alabiaga) [inline examples/amp-list.amp.html]: Iteration syntax with alternative custom delimiter is not that pleasant to look at. Maybe reconsider not using `#`?

File: examples/amp-list.amp.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (alabiaga) [reply examples/amp-list.amp.html]: done. good suggestion

**Pull request closed without merging**